crates/extensions/c8y_mapper_ext/src/operations/config_update.rs

```diff
@@ -91,7 +105,7 @@ impl CumulocityConverter {
             }
         };
 
-        Ok(messages)
+        Ok((messages, Some(command.into())))
```

Comment:
All the command-processing method are built along the same pattern:
- The method is given an MQTT `message`
- This `message` is parsed into a `command` of the appropriate type
- The `command` is processed leading a vector of MQTT `messages`
- The `command` is then converted into a `GenericCommand` to gather all log related data.
- The intermediate `command` acts a vehicle the `log_path` field without needing it.

So why not:

```suggestion
        Ok((messages, message.into()))
```

with an `impl From<MqttMessage> for Option<GenericCommand>` ?

I agree that this implies the message payload to be parsed twice, but this also keeps the command structs free of any unused `log_path`.

Reply:
> message payload to be parsed twice

This was the very first solution that came to my mind as well, but felt like an overkill and tried to avoid, as it's adding runtime cost to the operation handling just to parse an extra field. Of course, the impact is minimal, considering the overall cost of an operation handling.

> this also keeps the command structs free of any unused `log_path`

Now I think I understand your perspective. Yes, that `log_path` field would not make any sense to someone purely looking at the command payload struct APIs without any context of the mapper or the agent. So yes, it is clearly an example of the impl details creeping into the API, which should be avoided

One alternate solution which might be less intrusive and symmetric would be to add an additional `Value` field to all the `CommandPayload` implementors to capture any additional fields in the payload other than the essential ones.  We can argue that it also represents the open-ness of the payload. Then, we can implement an `impl<T: CommandPayload> From<Command<T>> for GenericCommandState` and eliminate the newly introduced `GenericCommand` struct. That way, we get symmetry as well, where the agent uses `GenericCommandState::set_log_path` and the converter uses the corresponding `GenericCommandState::get_log_path`, as you had suggested in an earlier comment. WDYT?

Reply:
> One alternate solution which might be less intrusive and symmetric would be to add an additional `Value` field to all the `CommandPayload` implementors to capture any additional fields in the payload other than the essential ones.  We can argue that it also represents the open-ness of the payload. Then, we can implement an `impl<T: CommandPayload> From<Command<T>> for GenericCommandState` and eliminate the newly introduced `GenericCommand` struct. That way, we get symmetry as well, where the agent uses `GenericCommandState::set_log_path` and the converter uses the corresponding `GenericCommandState::get_log_path`, as you had suggested in an earlier comment. WDYT?

Indeed, it's a good idea to add a `Value` field that holds the whole message payload with all the extra fields such as the `log_path` which are not used by the command itself but are required for bookkeeping.

The next question is where to store this `Value`? In all the `CommandPayload` implementations (as you suggest), or directly in the `Command<T>` struct along the decoded payload of type `T`? The latter might be less intrusive.

Reply:
> The next question is where to store this `Value`? In all the `CommandPayload` implementations (as you suggest), or directly in the `Command<T>` struct along the decoded payload of type `T`? The latter might be less intrusive.

I was thinking of keeping them in all the impls as that better represents the flexibility in the palyoad. Keeping another copy of `Value` with all the payload fields in the `Command` would have been easier to manage from the users' perspective, but I'm afraid that it'd again involves a 2-step parsing of the payload, once into `Value` and then into the concrete payload structs, unless there's some parsing trick that can deserialize the same byte stream into both the representations simultaneously.

Reply:
Unfortunately, the plan to have an `additional_fields: HashMap<String, Value>` field in all payloads to capture all non-essential fields did not quite work due to a conflict with the `CommandStatus` definition that has `serde(flatten)` attribute defined on it (primarily to capture the failed `status` and `reason` fields into the `Failed` same enum variant). When both `status` and `additional_fields` has `serde(flatten)` attribute defined on them, it was leading to a conflict where the `status` and `reason` fields get duplicated inside the `additional_fields` map as well. There was no easy way to fix this other than writing a custom deserialization logic, which would have complicated things for little value.

So, I was out of other options but to switch to the simpler solution of just parsing the `MqttMessage` into `GenericCommandState`, accepting the cost of double parsing.

Reply:
We decided to go with an explicit `log_path` added to all command specific payload. There are opportunities for simplifications, however not before https://github.com/thin-edge/thin-edge.io/pull/2844 is merged: one needs first to use the same representation for all commands. Currently, there are some small but blocking differences for Log & Config management commands.